akka-stream/src/main/scala/akka/stream/impl/Sources.scala

```diff
@@ -185,3 +191,143 @@ private[akka] final class SourceQueueAdapter[T](delegate: SourceQueueWithComplet
+            failStage(ex)
+          case Supervision.Restart ⇒
+            restartState()
+            resumingMode = true
+          case Supervision.Resume ⇒
+            resumingMode = true
+        }
+      }
+      if (resumingMode) onPull()
+    }
+
+    override def onDownstreamFinish(): Unit = closeStage()
+
+    private def restartState(): Unit = {
+      close(blockingStream)
+      blockingStream = create()
+    }
+
+    private def closeStage(): Unit =
+      try {
+        close(blockingStream)
+        completeStage()
+      } catch {
+        case NonFatal(ex) ⇒ failStage(ex)
+      }
+
+  }
+  override def toString = "UnfoldResourceSource"
+}
+
+private[stream] final class UnfoldResourceSourceAsync[T, S](create: () ⇒ Future[S],
+                                                            readData: (S) ⇒ Future[Option[T]],
+                                                            close: (S) ⇒ Future[Done]) extends GraphStage[SourceShape[T]] {
+  val out = Outlet[T]("UnfoldResourceSourceAsync.out")
+  override val shape = SourceShape(out)
+  override def initialAttributes: Attributes = DefaultAttributes.unfoldResourceSourceAsync
+
+  def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) with OutHandler {
+    lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+    var resource = Promise[S]()
+    implicit val context = ExecutionContexts.sameThreadExecutionContext
+
+    setHandler(out, this)
+
+    override def preStart(): Unit = createStream(false)
+
+    private def createStream(withPull: Boolean): Unit = {
+      val cb = getAsyncCallback[Try[S]] {
+        case scala.util.Success(res) ⇒
+          resource.success(res)
+          if (withPull) onPull()
+        case scala.util.Failure(t) ⇒ failStage(t)
+      }.invoke _
+      try {
+        create().onComplete(cb)
+      } catch {
+        case NonFatal(ex) ⇒ failStage(ex)
+      }
+    }
+
+    private def onResourceReady(f: (S) ⇒ Unit): Unit = resource.future.onSuccess {
+      case resource ⇒ f(resource)
+    }
+
+    val errorHandler: PartialFunction[Throwable, Unit] = {
+      case NonFatal(ex) ⇒ decider(ex) match {
+        case Supervision.Stop ⇒
+          onResourceReady(close(_))
+          failStage(ex)
+        case Supervision.Restart ⇒ restartState()
+        case Supervision.Resume  ⇒ onPull()
+      }
+    }
+    val callback = getAsyncCallback[Try[Option[T]]] {
+      case scala.util.Success(data) ⇒ data match {
+        case Some(d) ⇒ push(out, d)
+        case None    ⇒ closeStage()
+      }
+      case scala.util.Failure(t) ⇒ errorHandler(t)
+    }.invoke _
+
+    final override def onPull(): Unit = onResourceReady {
+      case resource ⇒
+        try { readData(resource).onComplete(callback) } catch errorHandler
+    }
+
+    override def onDownstreamFinish(): Unit = closeStage()
+
+    private def closeAndThen(f: () ⇒ Unit): Unit = {
```

Comment:
Shouldn't you call `setKeepGoing(true)` to make this work? Otherwise this stage stops on onDownstreamFinish.